Xbox extender stability fixes

paul-777 opened the pull request:
Fixes for xbox extender stability issues.

paul-777 committed:
Add motion to item select

Add motion to item selection to give early feedback on extenders where
page transitions can be very slow

paul-777 committed:
Ensure overscan is applied properly

Overscan settings were not being applied to some pages

paul-777 committed:
NullReferenceException in TVHelper

NullReferenceException in TVHelper when SeasonId and/or Series was null

paul-777 committed:
Threading stability changes

Standardise access to AddinHost and sub-objects to try and ensure
correct threading
Async was a pool of threadpools each of size one, combine some into a
pool which is allowed to grow larger. Async also allowed arbitrary names
for the pools that was allowing effective thread leaks when item names
where used as part of the pool name.
Accessing UI from non-UI threads was causing stability issues on xbox.
This was especially noticable when media was streaming which now appears
stable.

paul-777 committed:
revert commenting out done for testing

ebr11 commented on MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs:
```diff
@@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args)
             CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id;
 
-            Async.Queue("BasePlaybackController OnProgress", () =>
+            Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () =>
```
What makes this code required to run on the UI thread?

paul-777 replied:
That isn't pushing things onto the UI thread, if fact the opposite, it is still  running it Async, the change here is just to use an enum rather than string to identify the threadpool - as people were making up threadpool names that contained item names etc causing thread leaks. Using an enum ensures we keep control over how many threads/pools we create and what they are used for.

 

From: Eric Reed [mailto:notifications@github.com] 
Sent: 17 March 2015 12:12
To: MediaBrowser/MediaBrowser.Classic
Cc: Paul777
Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268)

 

In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26567168> :

> @@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args)
>              // Set the current PlayableItem based on the incoming args
>              CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id;
>  
> -            Async.Queue("BasePlaybackController OnProgress", () =>
> +            Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () =>

What makes this code required to run on the UI thread?

—
Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26567168> .  <https://github.com/notifications/beacon/AK4DMkEhcy-Qc6daaZvvUVjCnSZTrV6bks5n2BGggaJpZM4Dvs8_.gif>

paul-777 replied:
P.S. if you look at the enum you will see all the names I found in use, each of which was creating its own threadpool of size 1, not really much of a pool more a set of dedicated threads one for each type of use! and the threads once created were persistent. There were also entries that included item names as part of the name for the pool making things even worse.

Ideally that enum would be collapsed properly at some point, but for now I have just made a function which translates most of them to a common pool. Some of the use of the pools actually take and keep threads for the life of the process - e.g. the sqllite ones which isn't a typical use of a threadpool either and is the reason why I have kept those as having their own pools.

 

From: Paul Haley [mailto:paul@haleynet.co.uk] 
Sent: 17 March 2015 12:55
To: 'MediaBrowser/MediaBrowser.Classic'; 'MediaBrowser/MediaBrowser.Classic'
Subject: RE: [MediaBrowser.Classic] Xbox extender stability fixes (#268)

 

That isn't pushing things onto the UI thread, if fact the opposite, it is still  running it Async, the change here is just to use an enum rather than string to identify the threadpool - as people were making up threadpool names that contained item names etc causing thread leaks. Using an enum ensures we keep control over how many threads/pools we create and what they are used for.

 

From: Eric Reed [mailto:notifications@github.com] 
Sent: 17 March 2015 12:12
To: MediaBrowser/MediaBrowser.Classic
Cc: Paul777
Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268)

 

In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26567168> :

> @@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args)
>              // Set the current PlayableItem based on the incoming args
>              CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id;
>  
> -            Async.Queue("BasePlaybackController OnProgress", () =>
> +            Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () =>

What makes this code required to run on the UI thread?

—
Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26567168> .  <https://github.com/notifications/beacon/AK4DMkEhcy-Qc6daaZvvUVjCnSZTrV6bks5n2BGggaJpZM4Dvs8_.gif>

ebr11 replied:
That's sounds good.  Have you tested with all the plug-ins to be sure we haven't created any problems?

 

From: Paul777 [mailto:notifications@github.com] 
Sent: Tuesday, March 17, 2015 9:15 AM
To: MediaBrowser/MediaBrowser.Classic
Cc: Eric Reed
Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268)

 

In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26570621> :

> @@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args)
>              // Set the current PlayableItem based on the incoming args
>              CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id;
>  
> -            Async.Queue("BasePlaybackController OnProgress", () =>
> +            Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () =>

P.S. if you look at the enum you will see all the names I found in use, each of which was creating its own threadpool of size 1, not really much of a pool more a set of dedicated threads one for each type of use! and the threads once created were persistent. There were also entries that included item names as part of the name for the pool making things even worse. Ideally that enum would be collapsed properly at some point, but for now I have just made a function which translates most of them to a common pool. Some of the use of the pools actually take and keep threads for the life of the process - e.g. the sqllite ones which isn't a typical use of a threadpool either and is the reason why I have kept those as having their own pools. From: Paul Haley [mailto:paul@haleynet.co.uk] Sent: 17 March 2015 12:55 To: 'MediaBrowser/MediaBrowser.Classic'; 'MediaBrowser/MediaBrowser.Classic' Subject: RE: [MediaBrowser.Classic] Xbox extender stability fixes (#268 <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268> ) That isn't pushing things onto the UI thread, if fact the opposite, it is still running it Async, the change here is just to use an enum rather than string to identify the threadpool - as people were making up threadpool names that contained item names etc causing thread leaks. Using an enum ensures we keep control over how many threads/pools we create and what they are used for. From: Eric Reed [mailto:notifications@github.com] Sent: 17 March 2015 12:12 To: MediaBrowser/MediaBrowser.Classic Cc: Paul777 Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268 <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268> ) In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26567168> :

@@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args) // Set the current PlayableItem based on the incoming args CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id; - Async.Queue("BasePlaybackController OnProgress", () => + Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () =>

What makes this code required to run on the UI thread? — Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26567168> . <https://github.com/notifications/beacon/AK4DMkEhcy-Qc6daaZvvUVjCnSZTrV6bks5n2BGggaJpZM4Dvs8_.gif>

—
Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26570621> .  <https://github.com/notifications/beacon/ACmeiV7HM0-93pO8OHgH4cHgkiJoxz6vks5n2CBmgaJpZM4Dvs8_.gif>

paul-777 replied:
I haven't tested with the plugins - to be honest I don't use any so wouldn't know where to start - but if any plugins use the Async class via the old string methods they will get routed to a pool called Legacy which can grow up to 10 threads, unless the name matches one of the enum values is which case that one is used. Previously they would have got a pool of one thread for each name they used, so other than if a plugin is passing lots of random names and creating lots of long lived threads that it doesn't yield back to the pool I wouldn't expect any issues.

 

From: Eric Reed [mailto:notifications@github.com] 
Sent: 17 March 2015 14:04
To: MediaBrowser/MediaBrowser.Classic
Cc: Paul777
Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268)

 

In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26574469> :

> @@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args)
>              // Set the current PlayableItem based on the incoming args
>              CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id;
>  
> -            Async.Queue("BasePlaybackController OnProgress", () =>
> +            Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () =>

That's sounds good. Have you tested with all the plug-ins to be sure we haven't created any problems? From: Paul777 [mailto:notifications@github.com] Sent: Tuesday, March 17, 2015 9:15 AM To: MediaBrowser/MediaBrowser.Classic Cc: Eric Reed Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268 <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268> ) In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26570621> :

@@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args) // Set the current PlayableItem based on the incoming args CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id; - Async.Queue("BasePlaybackController OnProgress", () => + Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () =>

P.S. if you look at the enum you will see all the names I found in use, each of which was creating its own threadpool of size 1, not really much of a pool more a set of dedicated threads one for each type of use! and the threads once created were persistent. There were also entries that included item names as part of the name for the pool making things even worse. Ideally that enum would be collapsed properly at some point, but for now I have just made a function which translates most of them to a common pool. Some of the use of the pools actually take and keep threads for the life of the process - e.g. the sqllite ones which isn't a typical use of a threadpool either and is the reason why I have kept those as having their own pools. From: Paul Haley [mailto:paul@haleynet.co.uk] Sent: 17 March 2015 12:55 To: 'MediaBrowser/MediaBrowser.Classic'; 'MediaBrowser/MediaBrowser.Classic' Subject: RE: [MediaBrowser.Classic] Xbox extender stability fixes (#268 <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268>  <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268> ) That isn't pushing things onto the UI thread, if fact the opposite, it is still running it Async, the change here is just to use an enum rather than string to identify the threadpool - as people were making up threadpool names that contained item names etc causing thread leaks. Using an enum ensures we keep control over how many threads/pools we create and what they are used for. From: Eric Reed [mailto:notifications@github.com] Sent: 17 March 2015 12:12 To: MediaBrowser/MediaBrowser.Classic Cc: Paul777 Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268 <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268>  <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268> ) In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26567168> : @@ -120,7 +120,7 @@ protected virtual void OnProgress(PlaybackStateEventArgs args) // Set the current PlayableItem based on the incoming args CurrentPlayableItemId = args.Item == null ? Guid.Empty : args.Item.Id; - Async.Queue("BasePlaybackController OnProgress", () => + Async.Queue(Async.ThreadPoolName.BasePlaybackControllerOnProgress, () => What makes this code required to run on the UI thread? — Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26567168> . <https://github.com/notifications/beacon/AK4DMkEhcy-Qc6daaZvvUVjCnSZTrV6bks5n2BGggaJpZM4Dvs8_.gif> — Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26570621> . <https://github.com/notifications/beacon/ACmeiV7HM0-93pO8OHgH4cHgkiJoxz6vks5n2CBmgaJpZM4Dvs8_.gif>

—
Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26574469> .  <https://github.com/notifications/beacon/AK4DMpNRJ4iLcb_TO4cHZXVdaVuBuKokks5n2CvZgaJpZM4Dvs8_.gif>

ebr11 commented on MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BaseModelItem.cs:
```diff
@@ -68,12 +68,9 @@ public class BaseModelItem : IModelItem, IDisposable {
             });
         }
 
```
Why not change the implementation of FirePropertyChanged to be sure it always runs on the UI thread (like you had  it when you changed the name but without changing the name)?  That way it cannot be called incorrectly and you don't have to change all of the existing calls to it...

paul-777 replied:
For our BaseModelItem this is exactly what happens, I haven't undone anything it that area just changed the names back - which unfortunately makes it confusing, the only time you should see an explicit  invoke of FirePropertyChanged is when it is calling the Microsoft.MediaCenter.UI.Model.ModelItem.FirePropertyChanged method for objects which don't inherit from our BaseModelItem (e.g. Information, Clock, MenuItem). The confusion is why I had renamed our methods (obviously not something we can do now I'm aware of the plugin issue) to help me find the incorrect ones as having ours named the same as the MS one made it hard to find the mistakes.

Doing an explicit search on when we use the Application.UIDeferredInvokeIfRequired(() => FirePropertyChanged pattern I have just found 2 cases where it wasn't necessary (one in FolderModel and one in Item) and I have committed a change for them - not that they will do any real harm as the invoke is only done if necessary so will still only occur once just at a different point in the stack. They are just cases where I hadn't spotted them to remove.
 
- Paul

 

From: Eric Reed [mailto:notifications@github.com] 
Sent: 17 March 2015 12:16
To: MediaBrowser/MediaBrowser.Classic
Cc: Paul777
Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268)

 

In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BaseModelItem.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26567368> :

> @@ -68,12 +68,9 @@ public class BaseModelItem : IModelItem, IDisposable {
>              });
>          }
>  

Why not change the implementation of FirePropertyChanged to be sure it always runs on the UI thread (like you had it when you changed the name but without changing the name)? That way it cannot be called incorrectly and you don't have to change all of the existing calls to it...

—
Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26567368> .  <https://github.com/notifications/beacon/AK4DMu2D-hpfWOI3Y09ghNTxB6nD33zJks5n2BKOgaJpZM4Dvs8_.gif>

paul-777 committed:
remove unnecessary  Application.UIDeferredInvokeIfRequired

remove unnecessary  Application.UIDeferredInvokeIfRequired

ebr11 commented on MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs:
```diff
@@ -491,27 +491,32 @@ public virtual void SkipToNextInCollection()
 
         private void SkipTimerExpired(object sender, EventArgs args)
         {
+            Application.UIDeferredInvokeIfRequired(()=>{
```
This line right here is invoking on the UI thread, no?

paul-777 replied:
Yes, because the code it runs accesses the MediaExperience object both to read the position (which we may get away with but seems to cause some issues sometimes) and then to actually do the seek. I tried it with just having the actual SeekInternal implementation do the invoke but it was still causing some issues.

 

From: Eric Reed [mailto:notifications@github.com] 
Sent: 17 March 2015 14:00
To: MediaBrowser/MediaBrowser.Classic
Cc: Paul777
Subject: Re: [MediaBrowser.Classic] Xbox extender stability fixes (#268)

 

In MediaBrowser3WMC/MediaBrowser.WMC/Code/ModelItems/BasePlaybackController.cs <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268#discussion_r26574099> :

> @@ -491,27 +491,32 @@ public virtual void SkipToNextInCollection()
>  
>          private void SkipTimerExpired(object sender, EventArgs args)
>          {
> +            Application.UIDeferredInvokeIfRequired(()=>{

This line right here is invoking on the UI thread, no?

—
Reply to this email directly or view it on GitHub <https://github.com/MediaBrowser/MediaBrowser.Classic/pull/268/files#r26574099> .  <https://github.com/notifications/beacon/AK4DMgd-uJWCQ9gJNeAKW4BQOCeHJP7Cks5n2CrWgaJpZM4Dvs8_.gif>

paul-777 committed:
Revert "Ensure overscan is applied properly"

This reverts commit 90f7b12b9f8f7a1b8ea2b8f564a0fde58381469c.

paul-777 committed:
Revert "Revert "Ensure overscan is applied properly""

This reverts commit 89f1f13b8238bb4437ee0458592bda0924ba2194.

paul-777 committed:
Overscan fixes

Change overscan fix implementation so as to not modify file that would
affect other themes

paul-777 committed:
Reduce loggnig

Reduce logging

paul-777 committed:
Reduce logging

Rreduce logging

paul-777 committed:
revert changes to DevInstall.cmd

paul-777 committed:
Provide background indicator to allow easy overscan adjustment

Provides light blue borders round config screen if overscan needs
adjusting, adjustment can then be done until the borders just disappear.
Previously instructions about the overscan fields talked about blue
borders but they were never obvious on the config page itself.

paul-777 committed:
Debug asserts not right

this can happen when a metadata refresh occurs either as a message from
the server session or a user hitting the rec button

paul-777 committed:
FirePropertyChange tidy up, move UpdateActualThumbSize to be Async

Tidy up the way we call FirePropertyChanged and move
UpdateActualThumbSize to be Async

paul-777 committed:
LibraryImage was loading images from disk when it wasn't necessary

LibraryImage was loading images from disk when we just asked for the
path to the cached image causing a performance hit

paul-777 committed:
Remove excessive image loading

Setting AsyncImageLoader.Size doesn't do anything as the Library image
has previously been modified to no longer perform resizing or caching of
images at specific sizes. This means that PrimaryImage and
SmallPrimaryImage are the same and changing PreferredImageSmallSize has
no actual affect on the image that is loaded only on the
SmallImageDistorted property. This change ensures we don't load/reload
the image more than needed as it was causing a performance hit.

paul-777 committed:
Revert "Remove excessive image loading"

This reverts commit 27ab3d9054f613dfea1e3c6ae91cd099f8bbc146.

paul-777 committed:
Excess call to GetLocalImagePath

Ensure we only call the right overload of GetLocalImagePath not both

paul-777 committed:
Re-enable caching of resized images

Re-enable caching of resized images, this used to be present in MB2 but
had been removed. The result was that extenders were very slow to load
certain views especially on wifi connections as images are sent to the
extenders raw and resized there to be displayed. Resizing to the right
size on the media center host dramatically improves the extender
experience. In my example a folder that used to take 20secs now takes
3secs to open.

paul-777 committed:
Increase animDelay

increase AnimDelay so that extenders can actually scroll smoothly rather
than getting stuck and then building up a buffer of unprocessed
key/remote events that kept scrolling even after then user release the
remote button

paul-777 committed:
Use small image in focus rectangle

Use the PreferredImageSmall image in the focus rectangle as the
PreferredImage can be slow to transfer when running on an extender

paul-777 committed:
Increase full details delay

Increase the full details delay slightly to prevent extenders trying to
load it too soon when attempting to scrolling across quickly. May want
to make the different based on whether we are on an extender or not if
it proves too slow when using the native media center host.

paul-777 committed:
Improved button push animation and triggering

Improved button push animation and triggering method for posterViewItem
and coverflow items. Now goes up and down with the up down of the
physical button the user presses.

paul-777 committed:
Resized image cache changes

Only resize images when using an extender.
Change cache filename structure for resized images to make it easier to
see them

paul-777 committed:
Implement cache cleaning

Run a periodic sweep that deletes cache images that have not been
touched for 30 days

paul-777 committed:
Config parameter per user for UseResizedImages

Make use of resized images configurable in user config file, defaults
true for extender and false for host PC, but actually has significant
benefit on hosts with older graphics cards such as my Radeon X300 in my
dev host.
Config parameter is currently not exposed anywhere in the UI.

paul-777 committed:
Cache cleaning

Integrate previous commit with existing cache cleaning that I hadn't
found
